My Site Dashboard: DomainRegistrationSource Refresh

zwarm opened the pull request:
Parent #15215

This PR changes the way `DomainRegistrationSource` gets/refreshes data through the use of MediatorLiveData.
This PR is part of a series of PRs being written for adding pull-to-refresh to the MySite tab.


### To test
1. Launch the app
2. Navigate to the My Site screen.
3. Switch to a site on a free plan.
4. Notice the domain registration item is not shown.
5. Switch to a site on any paid plan and with unredeemed domain credit.
6. Notice the domain registration item is now being shown.
7. Tap on the domain registration item and notice the Domain Registration screen.
8. Register a domain name.
9. Back on the My Site screen, notice the domain registration item is not shown anymore.

## Regression Notes
1. Potential unintended areas of impact
Domain registration card doesn't show as expected

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing + automated tests

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Refactor: use mediatorLiveData for getting and refreshing data

zwarm committed:
Update failing tests

ParaskP7 reviewed (approved):
👋 @zwarm !

I have reviewed and tested this PR as per the instructions, everything works as expected, good job! 🌟

As per the discussion in #15554, have only [this](https://github.com/wordpress-mobile/WordPress-Android/pull/15554#discussion_r747407356) one suggestion (💡) for you to consider (*). I am going to approve this PR anyway, since none is blocking. I am NOT going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

(*) I am going to provide you with a refactoring patch for your convenience.

zwarm committed:
Refactor to use postState methodology

zwarm commented:
@ParaskP7 - Per our discussions, applied the patch and then adjusted with the following:
- implemented `postState` fun
- Moved the check for `shouldFetchPlans` into the general if
- Combined `null` and `true` branches in the when logic

🚀
